src/javascript/cons_list.ts

```diff
@@ -0,0 +1,82 @@
+// A type-safe immutable homogenous "persistent" cons list.
+// A cons list is either empty or a cons pair of T and a cons list.
+export type ConsList<T> = Cons<T>|Empty;
+
+// Creates a non-empty cons list.
+export function cons<T>(head: T, tail: ConsList<T> = empty()): Cons<T> {
```

Comment:
That's a matter of personal preference but cons(x) for a list of single element would look really weird. Maybe have a separate function singleton(x) if you have a lot of use cases for it?

Reply:
Good point, I've removed the default parameter and simply replaced the calls with explicit `cons(head, empty())` (there were only 4 of them).